### PR TITLE
add numpy 2 support

### PR DESCRIPTION
**Pull request by MuellerSeb**
Closes: #337 #345

Numpy 2.0.0rc1 was released: https://pypi.org/project/numpy/2.0.0c1 and is officially usable to build extensions.

- numpy 2.0.0rc1 for building extensions (for Python 3.9 and above)
- check multiple numpy and scipy versions in CI
- fixed minimal versions for numpy
- use `np.asarray` everywhere with `np.atleast_(n)d`
- fix long/longlong integer issue in cython on windows by always using 64bit integers

Additional doc fixes
- build docs with latest sphinx version
- require pyvista 0.40 at least

32bit builds were removed with this PR. I think we can happily drop that.

### CODE REVIEW

**Comment by MuellerSeb** [comment]
- Building with numpy 2 works :tada:
- Running tests with numpy 2 fails :sob:

A lot of the tests fail because of emcee and this issue: https://github.com/dfm/emcee/issues/509

Other errors have a problem with these lines in `estimator.pyx`:
```cython
    cdef long[:, :] counts = np.zeros((d_max, len(bin_edges)-1), dtype=long)
    cdef long[:] counts = np.zeros(len(bin_edges)-1, dtype=long)
    cdef long[:] counts = np.zeros(k_max, dtype=long)
    cdef long[:] counts = np.zeros(k_max, dtype=long)
```
and raise
```
E   ValueError: Buffer dtype mismatch, expected 'long' but got 'long long'
```

One other category of errors is:
```
If using `np.array(obj, copy=False)` replace it with `np.asarray(obj)` to allow a copy when needed (no behavior change in NumPy 1.x).
```
`np.array(..., copy=False)` is used in [krige/base.py](https://github.com/GeoStat-Framework/GSTools/blob/90003fbbe193cfa835a891b403b4024bb21867f9/src/gstools/krige/base.py#L387), 3 times in [variogram/variogram.py](https://github.com/GeoStat-Framework/GSTools/blob/90003fbbe193cfa835a891b403b4024bb21867f9/src/gstools/variogram/variogram.py#L265) and 3 times in [tools/geometric.py](https://github.com/GeoStat-Framework/GSTools/blob/90003fbbe193cfa835a891b403b4024bb21867f9/src/gstools/tools/geometric.py#L386)

**Comment by MuellerSeb** [comment]
All issues fixed. Only remaining issues should be solved upstream in emcee with: https://github.com/dfm/emcee/pull/510

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -82,36 +80,51 @@ jobs:
...
-        python-version: ['3.8', '3.9', '3.10', '3.11', '3.12']
-
+        # https://github.com/scipy/oldest-supported-numpy/blob/main/setup.cfg
+        ver:
+          - {py: '3.8', np: '==1.20.0', sp: '==1.5.4'}
+          - {py: '3.9', np: '==1.20.0', sp: '==1.5.4'}
+          - {py: '3.10', np: '==1.21.6', sp: '==1.7.2'}
+          - {py: '3.11', np: '==1.23.2', sp: '==1.9.2'}
+          - {py: '3.12', np: '==1.26.2', sp: '==1.11.2'}
+          - {py: '3.12', np: '>=2.0.0rc1', sp: '>=1.13.0'}
```
[the remark this answers is not in the document]
Comment (MuellerSeb) [reply .github/workflows/main.yml:96]: What exactly is awful? I think it is pretty neat to test multiple numpy/scipy versions to fix stuff like https://github.com/GeoStat-Framework/GSTools/issues/246 faster in the future.

File: pyproject.toml [hunk 2]
```diff
@@ -60,20 +61,20 @@ doc = [
...
 plotting = [
     "matplotlib>=3",
-    "pyvista>=0.29",
+    "pyvista>=0.40",
 ]
 rust = ["gstools_core>=0.2.0,<1"]
 test = ["pytest-cov>=3"]
 lint = [
-    "black",
+    "black>=23,<24",
```
[the remark this answers is not in the document]
Comment (MuellerSeb) [reply pyproject.toml:77]: black 24 introduced a blank line under the file doc-string. Without the cap, I would update almost all files to add a line there. And since black updates there policy changes once a year, this cap is reasonable.

File: src/gstools/variogram/estimator.pyx [hunk 3]
```diff
@@ -356,7 +357,7 @@ def ma_structured(
     cdef int k_max = i_max + 1
 
     cdef double[:] variogram = np.zeros(k_max)
-    cdef long[:] counts = np.zeros(k_max, dtype=long)
+    cdef int64_t[:] counts = np.zeros(k_max, dtype=np.int64)
```
[the remark this answers is not in the document]
Comment (MuellerSeb) [reply src/gstools/variogram/estimator.pyx:360]: I linked the issue above. Default integer behavior is changing with numpy 2 on windows. So being explicit here is better IMHO: https://numpy.org/devdocs/numpy_2_0_migration_guide.html#windows-default-integer

File: .github/workflows/main.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (LSchueler) [reply .github/workflows/main.yml:96]: It's just a lot of manual (and typo prone?) work. Just imagine we have to add another few packages to that version matrix.

And it also makes it more work to locally reproduce specific setups of users.

File: src/gstools/variogram/estimator.pyx (hunk 3 above)
[the remark this answers is not in the document]
Comment (LSchueler) [reply src/gstools/variogram/estimator.pyx:360]: I still don't see the problem, if we use `cdef int` instead of `cdef int64_t`. Does that fail on Windows?

[the remark this answers is not in the document]
Comment (MuellerSeb) [reply src/gstools/variogram/estimator.pyx:360]: I tried it, and it failed, yes: https://github.com/GeoStat-Framework/GSTools/pull/340/commits/4de1557881731c9a66c2ddea1d4a10b88a4f350c

File: .github/workflows/main.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (MuellerSeb) [reply .github/workflows/main.yml:96]: If a new Python version is release we have to add it by hand either way. And then I would just add the first version of numpy and scipy that support that new python version. That is how I selected the numpy and scipy versions for the python version together with a bleeding edge job for the latest versions.

File: pyproject.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (MuellerSeb) [reply pyproject.toml:77]: yes, I guess. So you would prefer configuring this instead of pinning the black version?

File: src/gstools/variogram/estimator.pyx (hunk 3 above)
[the remark this answers is not in the document]
Comment (MuellerSeb) [reply src/gstools/variogram/estimator.pyx:360]:
```
ValueError: Buffer dtype mismatch, expected 'int' but got 'long'
```

[the remark this answers is not in the document]
Comment (MuellerSeb) [reply src/gstools/variogram/estimator.pyx:360]: Do you think `int64_t` is worse in performance?

[the remark this answers is not in the document]
Comment (LSchueler) [reply src/gstools/variogram/estimator.pyx:360]: I'm pretty sure that the performance is the same. I would just like to not have to import some c files only to support a standard type.
This is likely not a problem, but take a look at [this](https://stackoverflow.com/a/62937389) for example.

Is that really the recommended way in Cython?

[the remark this answers is not in the document]
Comment (MuellerSeb) [reply src/gstools/variogram/estimator.pyx:360]: I don't know if it is the recommended way. What would you do?

File: pyproject.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (MuellerSeb) [reply pyproject.toml:77]: It seems, this is not configurable since it was standardized with black 24: https://github.com/psf/black/issues/4042

We could update to 24 (add blank line) and add a preview job:
```
python -m black --preview --diff --color .
```
That will show diffs from upcoming conventions but wont raise an error.

File: src/gstools/variogram/estimator.pyx (hunk 3 above)
[the remark this answers is not in the document]
Comment (MuellerSeb) [reply src/gstools/variogram/estimator.pyx:360]: Would you prefere `cnp.int64_t` from `cimport numpy as cnp`?

File: pyproject.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (LSchueler) [reply pyproject.toml:77]: Damn conformists ;-)

I mean, we don't forward the dependency to the user, so if this really is so complicated, I say do what's easiest for you.

File: src/gstools/variogram/estimator.pyx (hunk 3 above)
[the remark this answers is not in the document]
Comment (MuellerSeb) [reply src/gstools/variogram/estimator.pyx]: Done.

File: src/gstools/covmodel/fit.py [hunk 4]
```diff
@@ -308,9 +309,11 @@ def _pre_init_guess(model, init_guess, mean_x=1.0, mean_y=1.0):
...
         init_guess.setdefault(
             opt,
-            default_arg_from_bounds(bnd[opt])
-            if default
-            else getattr(model, opt),
+            (
+                default_arg_from_bounds(bnd[opt])
+                if default
+                else getattr(model, opt)
+            ),
```
Comment (MuellerSeb) [inline src/gstools/covmodel/fit.py:316]: This is the only black 24 change that is not related to the blank line below the module docstring.

File: pyproject.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (MuellerSeb) [reply pyproject.toml]: Applied black 24 and added a preview job.

File: .github/workflows/main.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (MuellerSeb) [reply .github/workflows/main.yml:98]: I would like to keep it like this for now. If we encounter version incompatibilities in the future, we can rethink this.

**Review by LSchueler — APPROVED** [review]
It'd be funny to see the number of changed files across open source  Python projects after changes in the black formats :-)

LGTM!
